src/metabase/models/dashboard_card.clj

```diff
@@ -217,6 +221,106 @@
     (events/publish-event! :dashboard-remove-cards {:id id :actor_id user-id :dashcards [dashboard-card]})))
 
+;;; ----------------------------------------------- Link cards ----------------------------------------------------
+
+(def ^:private all-card-info-columns
+  {:model         :text
+   :id            :integer
+   :name          :text
+   :description   :text
+
+   ;; for cards and datasets
+   :collection_id :integer
+   :display       :text
+
+   ;; for tables
+   :db_id        :integer})
+
+(def ^:private  link-card-columns-for-model*
+  {:database  #{:id :name :description}
+   :table     #{:id :name :description :db_id}
+   :dashboard #{:id :name :description :collection_id}
+   :card      #{:id :name :description :collection_id :display}
+   :dataset   #{:id :name :description :collection_id :display}})
+
+(defn- link-card-columns-for-model
+  [model]
+  (let [model-cols-set (link-card-columns-for-model* model)]
+    (for [[col col-type] all-card-info-columns]
+      (cond
+        (= col :model)
+        [(h2x/literal model) :model]
+
+        (model-cols-set col)
+        col
+        ;; This entity is missing the column, project a null for that column value. For Postgres and H2, cast it to the
+        ;; correct type, e.g.
+        ;;
+        ;;    SELECT cast(NULL AS integer)
+        ;;
+        ;; For MySQL, this is not needed.
+        :else
+        [(when-not (= (mdb/db-type) :mysql)
+           [:cast nil col-type])
+         col]))))
+
+(def ^:private link-card-model->toucan-model
+  {:database  :metabase.models.database/Database
+   :table     :metabase.models.table/Table
+   :dashboard :metabase.models.dashboard/Dashboard
+   :card      :metabase.models.card/Card
+   :dataset   :metabase.models.card/Card})
+
+(defn- link-card-info-query-for-model
+  [[model ids]]
+  {:select (link-card-columns-for-model model)
+   :from   (t2/table-name (link-card-model->toucan-model model))
+   :where  [:in :id ids]})
+
+(defn- link-card-info-query
+  [link-card-model->ids]
+  (if (= 1 (count link-card-model->ids))
+    (link-card-info-query-for-model (first link-card-model->ids))
+    {:select   [:*]
+     :from     [[{:union-all (map link-card-info-query-for-model link-card-model->ids)}
+                 :alias_is_required_by_sql_but_not_needed_here]]}))
+
+(def ^:private link-card-models
+  #{"database" "table" "dataset" "card" "dashboard"})
+
+(defn with-link-card-info
+  "Update entity info for link cards.
+
+  Link cards are dashcards that link to internal entities like Database/Dashboard/... or an url.
+  The viz-settings only store the model name and id, info like name, description will need to be
+  hydrated on fetch to make sure those info are up-to-date."
+  [dashcards]
+  (let [entity-path        [:visualization_settings :link :entity]
+        ;; find all dashcards that are link-cards and get its model, id
+        ;; [[:table #{1 2}] [:database #{3 4}]]
+        model-and-ids      (->> dashcards
```

Comment:
this is a bit complicated because we want to query all the entities at once for a dashboard instead of doing multiple roundtrips to the DBS.